src/PersistentOrderedMap.mo

```diff
@@ -0,0 +1,925 @@
+    }
+  };
+
+  /// Collapses the elements in `rbMap` into a single value by starting with `base`
+  /// and progressively combining keys and values into `base` with `combine`. Iteration runs
+  /// right to left.
+  ///
+  /// Example:
+  /// ```motoko
+  /// import Map "mo:base/PersistentOrderedMap";
+  /// import Nat "mo:base/Nat"
+  /// import Iter "mo:base/Iter"
+  ///
+  /// let mapOps = Map.MapOps<Nat>(Nat.compare);
+  /// let rbMap = mapOps.fromIter<Text>(Iter.fromArray([(0, "Zero"), (2, "Two"), (1, "One")]));
+  ///
+  /// func folder(key : Nat, val : Text, accum : (Nat, Text)) : ((Nat, Text))
+  ///   = (key + accum.0, accum.1 # val);
+  ///
+  /// Debug.print(debug_show(Map.foldRight(rbMap, (0, ""), folder)));
+  ///
+  /// // (3, "TwoOneZero")
+  /// ```
+  ///
+  /// Cost of iteration over all elements:
+  /// Runtime: `O(n)`.
+  /// Space: depends on `combine` function plus garbage, see the note below.
+  /// where `n` denotes the number of key-value entries stored in the map.
+  ///
+  /// Note: Full map iteration creates `O(n)` temporary objects that will be collected as garbage.
+  public func foldRight<Key, Value, Accum>(
+    rbMap : Map<Key, Value>,
+    base : Accum,
+    combine : (Key, Value, Accum) -> Accum
+  ) : Accum
+  {
+    switch (rbMap) {
+      case (#leaf) { base };
+      case (#red(l, (k, v), r)) {
+        let right = foldRight(r, base, combine);
+        let middle = combine(k, v, right);
+        foldRight(l, middle, combine)
+      };
+      case (#black(l, (k, v), r)) {
+        let right = foldRight(r, base, combine);
+        let middle = combine(k, v, right);
+        foldRight(l, middle, combine)
+      }
+    }
+  };
+
+
+  public module Internal {
+
+    public func fromIter<K, V>(i : I.Iter<(K,V)>, compare : (K, K) -> O.Order) : Map<K, V>
+    {
+      var map = #leaf : Map<K,V>;
+      for(val in i) {
+        map := put(map, compare, val.0, val.1);
+      };
+      map
+    };
+
+    public func mapFilter<K, V1, V2>(t : Map<K, V1>, compare : (K, K) -> O.Order, f : (K, V1) -> ?V2) : Map<K, V2>{
+      func combine(key : K, value1 : V1, acc : Map<K, V2>) : Map<K, V2> {
+        switch (f(key, value1)){
+          case null { acc };
+          case (?value2) {
+            put(acc, compare, key, value2)
+          }
+        }
+      };
+      foldLeft(t, #leaf, combine)
+    };
+
+    public func get<K, V>(t : Map<K, V>, compare : (K, K) -> O.Order, x : K) : ?V {
+      switch t {
+        case (#leaf) { null };
+        case (#red(l, xy, r)) {
+          switch (compare(x, xy.0)) {
+            case (#less) { get(l, compare, x) };
+            case (#equal) { ?xy.1 };
+            case (#greater) { get(r, compare, x) }
+          }
+        };
+        case (#black(l, xy, r)) {
+          switch (compare(x, xy.0)) {
+            case (#less) { get(l, compare, x) };
+            case (#equal) { ?xy.1 };
+            case (#greater) { get(r, compare, x) }
+          }
+        }
+      }
+    };
```

Comment:
Since get has regressed with inlining the colours (more cases, I guess) maybe it would be worth re-ordering the patterns so leaves are checked last (and with just a default pattern). If that helps get, it might help in other similar code  too.

Of course, it would be nice if the compiler this for you (but it does not as far as I know (right @ggreif?)).

```suggestion
    public func get<K, V>(t : Map<K, V>, compare : (K, K) -> O.Order, x : K) : ?V {
      switch t {
        case (#red(l, xy, r)) {
          switch (compare(x, xy.0)) {
            case (#less) { get(l, compare, x) };
            case (#greater) { get(r, compare, x) }
            case _ { ?xy.1 };
          }
        };
        case (#black(l, xy, r)) {
          switch (compare(x, xy.0)) {
            case (#less) { get(l, compare, x) };
            case (#greater) { get(r, compare, x) }
            case _ { ?xy.1 };
          }
        };
        case _ { null };
      }
    };
```

Reply:
Perhaps that's because folds and maps need to visit all the leaves, while put and get only need to consider one?
Another tweak would be to use a wildcard pattern for the last default case (when applicable), which would (I expect) avoid a test, i.e.
 
 ```
 //last case
case #leaf { ... }  -> 
````
```
//last case
 case _ { ... }
```

Reply:
Thanks for doing the experiment!

Reply:
Yeah, I think I'd agree.

Reply:
We've merged this optimization and updated the description